signing_service/src/signing_service.py

```diff
@@ -0,0 +1,151 @@
+            self.is_sigterm_sent = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_is_sigterm_sent_true)
+
+        while True:
+            if self.is_sigterm_sent:
+                self.socket.close()
+                logger.info('Closing connection and exiting on SIGTERM.')
+                break
+            try:
+                if self.socket is None:
+                    self._connect()
+                else:
+                    self.socket.send(b'')
+            except (socket.error, socket.gaierror) as exception:  # TODO: Test
+                logger.error(f'Socket error occurred: {exception}')
+
+                # When the other side closes the connection, return without errors
+                if isinstance(exception.args, tuple) and exception.args[0] == socket.errno.EPIPE:  # type: ignore
+                    self.socket.close()
+                    logger.info(os.strerror(exception.args[0]))
+                    break
+
+                # Increase delay and reconnect if the connection is interrupted due to a failure.
+                self._increase_delay()
+                self.socket = None
+            except KeyboardInterrupt:
+                # Handle keyboard interrupt.
+                self.socket.close()
+                logger.info('Closing connection and exiting on KeyboardInterrupt.')
+                break
+            except Exception as exception:
+                # Try to close opened socket.
+                self.socket.close()
+                # If there was an unrecognized exception, log it and report to Sentry.
+                logger.error(f'Unrecognized exception occurred: {exception}')
+                crash_logger.error(str(exception))
+                raise
+
+    def _connect(self) -> None:
+        """ Creates socket and connects to given HOST and PORT. """
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        if self.current_reconnect_delay > 0:
+            logger.warning(f'Waiting {self.current_reconnect_delay} before connecting.')
+            sleep(self.current_reconnect_delay)
+
+        logger.warning(f'Connecting to {self.host}:{self.port}.')
+        self.socket.connect((self.host, self.port))
+        logger.warning(f'Connection established.')
+
+        # Reset delay on successful connection and set flag that connection is established.
+        self.current_reconnect_delay = 0
+        self._handle_connection()
+
+    def _handle_connection(self) -> None:
+        """ Inner loop that handles data exchange over socket. """
+        while True:
+            data = self.socket.recv(1024)
+            self.socket.send(data)
+
+    def _increase_delay(self) -> None:
+        """ Increase current delay if connection cannot be established. """
+        if self.current_reconnect_delay == 0:
+            self.current_reconnect_delay = self.reconnect_delay
```

Comment:
I think that `None` would be a better initial value for `current_reconnect_delay` because trying to multiply it by 2 by mistake results in an error.

And I'd assert that it's never zero.

Reply:
:+1: done